### PR TITLE
Fix/better user lib for authentication

### PR DESCRIPTION
**Pull request by inkhey**
Depends on #56.
-> Remove swagger_ui view
-> Allow in_context params in many user_lib method (is it the best way to deal with this problem) to return UserInContext object.
-> Add autehnticate_user method in user_lib
-> Add tests+renaming for user_lib.

### CODE REVIEW

File: tracim/lib/core/user.py [hunk 1]
```diff
@@ -14,20 +18,79 @@ def __init__(self, current_user: typing.Optional[User], session, config):
...
+        if not self._user:
+            raise UserNotExist()
+        return self._user
+
+    def get_all(self) -> typing.Iterable[User]:
+        return self._session.query(User).order_by(User.display_name).all()
+
+    # Check methods
+
+    def user_with_email_exists(self, email: str):
```
Comment (lebouquetin) [inline tracim/lib/core/user.py]: missing return typing

File: tracim/lib/core/user.py [hunk 2]
```diff
@@ -14,20 +18,79 @@ def __init__(self, current_user: typing.Optional[User], session, config):
...
+        :param email: email of the user
+        :param password: cleartext password of the user
+        :return: User who was authenticated.
+        """
+        try:
+            user = self.get_one_by_email(email)
+            if user.validate_password(password):
+                return user
+            else:
+                raise BadUserPassword()
```
Comment (lebouquetin) [inline tracim/lib/core/user.py]: why do you return an instance of exception here and a class 3 lines further?

File: tracim/tests/library/test_user_api.py [hunk 3]
```diff
@@ -60,26 +63,84 @@ def test_get_one_by_email_exception(self):
...
+    def test_unit__get_user_with_context__nominal_case(self):
+        user = User(email='admin@tracim.tracim')
+        api = UserApi(
+            current_user=None,
+            session=self.session,
+            config=self.config,
+        )
+        new_user = api.get_user_with_context(user)
+        assert isinstance(new_user, UserInContext)
+        assert new_user.user == user
```
Comment (lebouquetin) [inline tracim/tests/library/test_user_api.py:102]: you should also check that the context is valid (eg: user avatar url, ...)

File: tracim/tests/library/test_user_api.py [hunk 4]
```diff
@@ -60,26 +63,84 @@ def test_get_one_by_email_exception(self):
...
+        api = UserApi(
+            current_user=None,
+            session=self.session,
+            config=self.config,
+        )
+        user = api.authenticate_user('admin@admin.admin', 'admin@admin.admin')
+        assert isinstance(user, User)
+        assert user.email == 'admin@admin.admin'
+
+    def test_unit__authenticate_user___err__bad_password(self):
```
Comment (lebouquetin) [inline tracim/tests/library/test_user_api.py]: replace bad password by wrong password

File: tracim/tests/library/test_user_api.py [hunk 5]
```diff
@@ -60,26 +63,84 @@ def test_get_one_by_email_exception(self):
...
+            api.authenticate_user('admin@admin.admin', 'bad_password')
+
+    def test_unit__authenticate_user___err__bad_user(self):
+        api = UserApi(
+            current_user=None,
+            session=self.session,
+            config=self.config,
+        )
+        with pytest.raises(AuthenticationFailed):
+            api.authenticate_user('unknown_user', 'bad_password')
```
Comment (lebouquetin) [inline tracim/tests/library/test_user_api.py]: bad password means "un mot de passe de mauvaise qualité". wrong password means "un mot de passe erroné". you should use WrongPassword (even rename the exception BadPassword to WrongPassword)

File: tracim/views/core_api/session_controller.py [hunk 6]
```diff
@@ -1,72 +1,47 @@
...
+    @hapic.output_body(NoContentSchema(), default_http_code=HTTPStatus.NO_CONTENT)  # nopep8
     def login(self, context, request: TracimRequest, hapic_data=None):
         """
         Logs user into the system
         """
-        email = request.params['email']
-        password = request.params['password']
-        if not (email and password):
-            raise Exception
+        email = request.json_body['email']
```
Comment (lebouquetin) [inline tracim/views/core_api/session_controller.py]: ask @buxx how to automatically fill an object from Schema in order not to have to fill yourself stuff from request.json_body.

File: tracim/views/core_api/session_controller.py [hunk 7]
```diff
@@ -1,72 +1,47 @@
...
-        except NoResultFound:
-            # User does not exist
-            raise LoginFailed('Bad Credentials')
-        return
+        uapi = UserApi(
+            None,
+            session=request.dbsession,
+            config=app_config,
+        )
+        return uapi.authenticate_user(email, password)
```
Comment (lebouquetin) [inline tracim/views/core_api/session_controller.py]: after the object is filled using Schema, the code here will be something like:

```
return uapi.authenticate_user(login.email, login.password)
```

File: tracim/views/core_api/session_controller.py [hunk 8]
```diff
@@ -75,54 +50,31 @@ def logout(self, context, request: TracimRequest, hapic_data=None):
...
-        configurator.add_route(
-            'login',
-            os.path.join(BASE_API_V2, 'sessions', 'login'),
-            request_method='GET'
-        )
-        configurator.add_view(
-            self.login,
-            route_name='login',
-        )
+        configurator.add_route('login', '/sessions/login', request_method='POST')  # nopep8
```
Comment (lebouquetin) [inline tracim/views/core_api/session_controller.py:71]: would be nice to defin routes directly as view method decorators. is it possible ?

**Review by lebouquetin — CHANGES_REQUESTED** [review]
@inkhey some changes + the subject about route definition as a decorator (maybe we should talk about this with @buxx )

File: tracim/views/core_api/session_controller.py (hunk 8 above)
Comment (inkhey) [reply tracim/views/core_api/session_controller.py]: complicated.

File: tracim/tests/library/test_user_api.py (hunk 3 above)
Comment (inkhey) [reply tracim/tests/library/test_user_api.py:102]: Done with commit  d9cdd657fbe6d6c5cb89d87e77091010b0f73ca6

File: tracim/views/core_api/session_controller.py (hunk 8 above)
Comment (inkhey) [reply tracim/views/core_api/session_controller.py:71]: we previously decide to not rely on pyramid default decorator :

It's not in pyramid philosophy to have decorator who create both view and route at the same time like in some others web framework, pyramid explicitly distinct view and route. That's why such type of decorator is not a good idea.

Working with default pyramid decorator who are only view decorator doesn't simplify a lot code as we already need to set route in another place .

Class method view decorator doesn't work well with Hapic , see https://github.com/algoo/hapic/issues/42

**Pull request merged**